rfcs/rfc-using-pinia-for-form-components.md

```diff
@@ -0,0 +1,318 @@
+- Start Date: 2023-03-06
+- Target Major Version: n/a
+- Reference Issues: bcgov/entity#15509
+- Entity Issue: 
+- Implementation PR: 
+
+# Summary
+
+This RFC seeks feedback to help improve the application architecture of
+business-filings-ui, business-create-ui and business-filings-ui.  At present,
+we use props and events to communicate between form components like this:
+
+```javascript
+<CourtOrderPoa
+  :draftCourtOrderNumber="getCourtOrderStep.courtOrder.fileNumber"
+  :hasDraftPlanOfArrangement="getCourtOrderStep.courtOrder.hasPlanOfArrangement"
+  :courtOrderNumberRequired="false"
+  :invalidSection="isCourtOrderInvalid"
+  @emitCourtNumber="setCourtOrderFileNumber(\$event)"
+  @emitPoa="setHasPlanOfArrangement(\$event)"
+  @emitValid="setCourtOrderValidity(\$event)"
+/>
+```
+
+We're proposing to modify the form components to use Pinia getters and mutators instead.
+
+# Basic example
+
+Each form component would be passed a store location as a string like this: 
+```javascript
+<CourtOrderPoa path="some.location.in.the.store" />
+```
```

Comment:
@cameron-eyds since individual fields (like input, text-area, checkbox, radio button, etc.) talk directly to the store, I see this giving us more flexibility.  I would suggest that the individual form fields become our primary shared components.  We would create higher-level shared components from those basic building blocks which could be easily to recreated when the composite component don't fit our requirement.

Reply:
Ah i see, thats pretty interesting! 
Handling the data at a bit more of a granular level, thank you for clarifying.

Reply:
There is a lot of value in the high-level common/shared components, that I wouldn't want to lose. They contain a lot of complexity, which is often reused, and that doesn't make sense to reinvent each time it's needed.

Reply:
Here's an example of a "high-level" shared component made up of individual fields:
```html
<FormRow>
  <TextField id="address" :path="path" placeholder="Address" rules="max:25">
      Address Line
  </TextField>
  <TextField id="city" :path=path rules="max:20">City</TextField>
  <ProvinceField id="province" :path=path>Prov / State</ProvinceField>
  <TextField id="postal" :path=path rules="postal">Postal / Zip</TextField>
  <PhoneField id="phone" :path=path rules="phone">Phone</PhoneField>
</FormRow>
```
Since the component includes only layout and validation rules, it's easy to create a new component that has an additional field or different validation logic.

Reply:
Cool, thanks.

Do you have a sample of a common component that uses resources for things like labels or rules?

Reply:
> There is a lot of value in the high-level common/shared components, that I wouldn't want to lose. They contain a lot of complexity, which is often reused, and that doesn't make sense to reinvent each time it's needed.

That's a good point @severinbeauvais I'd like to better understand the complexity in shared components.  My gut feeling is that when shared components include a lot of complexity, they become more difficult to customize for a specific use case.

Reply:
> My gut feeling is that when shared components include a lot of complexity, they become more difficult to customize for a specific use case.

It depends on the use cases. That complexity has to be somewhere, and if it's the same logic for multiple filings/apps then I think it makes sense to encapsulate it. If you build a car, you don't want to have to redesign a new transmission -- you just want to use an off-the-shelf one.

Reply:
> Do you have a sample of a common component that uses resources for things like labels or rules?

I think that would be straight forward.  Using the example above, replace the labels and rules with dynamic values that can be either passed in as a prop or retrieved from the store using a getter.

```javascript
<FormRow>
  <TextField id="address" :path="path" placeholder="Address" :rules="address.rules">
      {{ address.label }}
  </TextField>
  <TextField id="city" :path=path :rules="city.rules">{{ city.label }}</TextField>
  <ProvinceField id="province" :path=path>{{ province.label }}</ProvinceField>
  <TextField id="postal" :path=path :rules="postal.rules">{{ postal.label }}</TextField>
  <PhoneField id="phone" :path=path :rules="phone.rules">{{ phone.label }}</PhoneField>
</FormRow>
```

Reply:
OK, cool.

So this is really just a different way for a component to get and set data (no more props to send in variable data or events to get out variable data).